rclcpp/CMakeLists.txt

```
@@ -221,6 +222,13 @@ if(BUILD_TESTING)
 
   add_definitions(-DTEST_RESOURCES_DIRECTORY="${CMAKE_CURRENT_BINARY_DIR}/test/resources")
 
+  rosidl_generate_interfaces(${PROJECT_NAME}_test_msgs
+    test/msg/Header.msg
+    test/msg/MessageWithHeader.msg
+    DEPENDENCIES builtin_interfaces
+    LIBRARY_NAME ${PROJECT_NAME}
```

Since we're only generating the messages for the tests in this package, we don't need to install them. We can add `SKIP_INSTALL`.
```suggestion
    LIBRARY_NAME ${PROJECT_NAME}
    SKIP_INSTALL
```

This should also eliminate the need to add the `<member_of_group>` tag.

---

When I add `SKIP_INSTALL` I get the same error as before:

```bash
Failed to find library 'rclcpp__rosidl_typesupport_fastrtps_cpp'
unknown file: Failure
C++ exception with description "could not create publisher: type support not from this implementation, at /Users/prajaktg/ros2_ws/src/ros2/rmw_fastrtps/rmw_fastrtps_cpp/src/publisher.cpp:75, at /Users/prajaktg/ros2_ws/src/ros2/rcl/rcl/src/rcl/publisher.c:173" thrown in the test body.
```

---

@dabonnie can you verify that with the latest changes on this branch? Trying to see if it's just my workspace doing that.

---

I can reproduce the same issue with your branch.. I'm not sure why skipping installation causes the issue :thinking:

---

In any case, I don't think it's a blocker. We could figure it out later and patch it after.

---

@prajakta-gokhale yes, I was able to build with all of @jacobperron's suggestions. Will push shortly.

---

@jacobperron here's my diff, are we missing something?

```
diff --git a/rclcpp/CMakeLists.txt b/rclcpp/CMakeLists.txt
index d57289c..68b8f5a 100644
--- a/rclcpp/CMakeLists.txt
+++ b/rclcpp/CMakeLists.txt
@@ -227,6 +227,7 @@ if(BUILD_TESTING)
     test/msg/MessageWithHeader.msg
     DEPENDENCIES builtin_interfaces
     LIBRARY_NAME ${PROJECT_NAME}
+    SKIP_INSTALL
   )
 
   ament_add_gtest(test_client test/test_client.cpp)
diff --git a/rclcpp/package.xml b/rclcpp/package.xml
index 9c4b6e2..4e7f15b 100644
--- a/rclcpp/package.xml
+++ b/rclcpp/package.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
-<package format="3">
+<package format="2">
   <name>rclcpp</name>
   <version>0.8.3</version>
   <description>The ROS client library in C++.</description>
@@ -40,8 +40,6 @@
   <test_depend>rosidl_default_generators</test_depend>
   <test_depend>test_msgs</test_depend>
 
-  <member_of_group>rosidl_interface_packages</member_of_group>
-
   <export>
     <build_type>ament_cmake</build_type>
   </export>
```

---

Ah, since we're not installing the package, we need to point the test executable to the right place.

Modifying the CMakeList.txt like so resolves it for me:

```diff
- ament_add_gtest(test_subscription_topic_statistics test/topic_statistics/test_subscription_topic_statistics.cpp)
+ ament_add_gtest(test_subscription_topic_statistics test/topic_statistics/test_subscription_topic_statistics.cpp
+   APPEND_LIBRARY_DIRS "${CMAKE_CURRENT_BINARY_DIR}"
+ )

---

@jacobperron still results in the same failure with a clean build :-(